File: EbookLibraryUI/Models/EbookDto.cs
using System.Text.Json.Serialization;

namespace EbookLibraryUI.Models;

/// <summary>Mirrors the EbookResponse schema returned by the FastAPI backend.</summary>
public class EbookDto
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; }

    [JsonPropertyName("title")]
    public string? Title { get; set; }

    [JsonPropertyName("isbn")]
    public string? Isbn { get; set; }

    [JsonPropertyName("authors")]
    public List<string> Authors { get; set; } = [];

    [JsonPropertyName("year")]
    public int? Year { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("category")]
    public string? Category { get; set; }

    [JsonPropertyName("subcategory")]
    public string? Subcategory { get; set; }

    [JsonPropertyName("publisher")]
    public string? Publisher { get; set; }

    [JsonPropertyName("edition")]
    public string? Edition { get; set; }

    [JsonPropertyName("language")]
    public string? Language { get; set; }

    [JsonPropertyName("page_count")]
    public int? PageCount { get; set; }

    [JsonPropertyName("file_name")]
    public string? FileName { get; set; }

    [JsonPropertyName("file_path")]
    public string? FilePath { get; set; }

    [JsonPropertyName("cover_image_path")]
    public string? CoverImagePath { get; set; }

    [JsonPropertyName("cover_image_mime_type")]
    public string? CoverImageMimeType { get; set; }

    [JsonPropertyName("has_errors")]
    public bool HasErrors { get; set; }

    [JsonPropertyName("is_metadata_stored")]
    public bool IsMetadataStored { get; set; }

    [JsonPropertyName("is_embeded_data_stored")]
    public bool IsEmbededDataStored { get; set; }

    /// <summary>Comma-separated authors string for display in table cells.</summary>
    public string AuthorsDisplay => Authors.Count > 0 ? string.Join(", ", Authors) : "—";

    /// <summary>Absolute local file URI for the cover image shown in WPF.</summary>
    public string? CoverUrl
    {
        get
        {
            if (string.IsNullOrWhiteSpace(CoverImagePath))
                return null;
            try
            {
                var full = System.IO.Path.GetFullPath(CoverImagePath.Trim());
                return System.IO.File.Exists(full) ? new Uri(full).AbsoluteUri : null;
            }
            catch (ArgumentException)
            {
                return null;
            }
            catch (NotSupportedException)
            {
                return null;
            }
        }
    }
}

File: EbookLibraryWeb/src/models/EbookDto.ts
const API_BASE = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000';

export interface EbookDto {
  id: string;
  title: string | null;
  isbn: string | null;
  authors: string[] | null;
  year: number | null;
  description: string | null;
  category: string | null;
  subcategory: string | null;
  publisher: string | null;
  edition: string | null;
  language: string | null;
  page_count: number | null;
  file_name: string | null;
  cover_image_path: string | null;
  cover_image_mime_type: string | null;
  has_errors: boolean;
  is_metadata_stored: boolean;
  is_embeded_data_stored: boolean;
}

export function getAuthorsDisplay(ebook: EbookDto): string {
  return ebook.authors?.join(', ') ?? '';
}

export function getCoverUrl(ebook: EbookDto): string | null {
  if (!ebook.cover_image_path) return null;
  return `${API_BASE}/covers/${ebook.cover_image_path}`;
}
